README.md

```diff
@@ -1,2 +1,174 @@
 # pytorch_cmspepr
+
 pytorch bindings for optimized knn and aggregation kernels
+
+
+## Example
+
+```python
+>>> import torch
+>>> import torch_cmspepr
+
+# Two events with 5 nodes and 4 nodes, respectively.
+# Nodes here are on a diagonal line in 2D, with d^2 = 0.02 between them.
+>>> nodes = torch.FloatTensor([
+    # Event 0
+    [.1, .1],
+    [.2, .2],
+    [.3, .3],
+    [.4, .4],
+    [100., 100.],
+    # Event 1
+    [.1, .1],
+    [.2, .2],
+    [.3, .3],
+    [.4, .4]
+    ])
+# Designate which nodes belong to which event
+>>> batch = torch.LongTensor([0,0,0,0,0,1,1,1,1])
+
+# Generate edges: k=2, max_radius^2 of 0.04
+>>> torch_cmspepr.knn_graph(nodes, 2, batch, max_radius=.2)
+tensor([[0, 1, 1, 2, 2, 3, 5, 6, 6, 7, 7, 8],
+        [1, 0, 2, 1, 3, 2, 6, 5, 7, 6, 8, 7]])
+
+# Generate edges: k=3 with loops allowed
+>>> torch_cmspepr.knn_graph(nodes, 3, batch, max_radius=.2, loop=True)
+tensor([[0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5, 5, 6, 6, 6, 7, 7, 7, 8, 8],
+        [0, 1, 1, 0, 2, 2, 1, 3, 3, 2, 4, 5, 6, 6, 5, 7, 7, 6, 8, 8, 7]])
+
+# If CUDA is available, the CUDA version of the knn_graph is used automatically:
+>>> gpu = torch.device('cuda') 
+>>> torch_cmspepr.knn_graph(nodes.to(gpu), 2, batch.to(gpu), max_radius=.2)
+tensor([[0, 1, 1, 2, 2, 3, 5, 6, 6, 7, 7, 8],
+        [1, 0, 2, 1, 3, 2, 6, 5, 7, 6, 8, 7]], device='cuda:0')
+```
+
+
+## Installation and requirements
+
+v1 is tested with CUDA 11.7 and pytorch 2.0.
+You should verify `nvcc` is available:
+
+```console
+$ nvcc --version
+nvcc: NVIDIA (R) Cuda compiler driver
+Copyright (c) 2005-2022 NVIDIA Corporation
+Built on Wed_Jun__8_16:49:14_PDT_2022
+Cuda compilation tools, release 11.7, V11.7.99
+Build cuda_11.7.r11.7/compiler.31442593_0
+```
+
+Also a `gcc` version of 5 or higher is recommended.
+
+The package is not (yet) available on PyPI, so local installation is at the moment the
+preferred installation method:
+
+```bash
+git clone git@github.com:cms-pepr/pytorch_cmspepr.git
+cd pytorch_cmspepr
+pip install -e .
+```
+
+Installing _only_ the CPU or CUDA extensions is supported:
+
+```bash
+FORCE_CPU_ONLY=1 pip install -e .  # Only compile C++ extensions
+FORCE_CUDA_ONLY=1 pip install -e .  # Only compile CUDA extenstions
+FORCE_CUDA=1 pip install -e .  # Try to compile CUDA extenstion even if no device found
+```
+
+If you only want to test the compilation of the extensions:
+
+```bash
+python setup.py develop
+```
+
+### Containerization
+
+It is recommended to install and run inside a container.
+At the time of writing (29 Sep 2023), the [pytorch/pytorch:2.0.0-cuda11.7-cudnn8-devel](https://hub.docker.com/layers/pytorch/pytorch/2.0.0-cuda11.7-cudnn8-devel/images/sha256-96ccb2997a131f2455d70fb78dbb284bafe4529aaf265e344bae932c8b32b2a4?context=explore)
```

Comment:
I set this up successfully on a machine with 1080s at MIT, but on another machine we currently only have drivers that support up to CUDA 11.6. I don't see any PyTorch dockers that support torch 2.0 and CUDA 11.6. Are both required?

Reply:
torch 2.0 strictly requires cuda >= 11.7

Reply:
Does this implementation require torch2.0? In any case, I was messing around with other containers and concluded this is too much of a pain, I'll use the 1080s for now until we can update the Nvidia drivers on the A30s

Reply:
Looking at the changes in this PR it doesn't look like wouldn't compile against torch < 2 and corresponding cuda.

Though, my two cents, it's probably just better to migrate your projects up to torch 2, it's pretty damn sweet featurewise.

Reply:
Yeah, I'll use the 1080s until the A30s are updated.

Reply:
For what it's worth, version 0.0.1 should work with torch 1.0... but it also needs CUDA 10 I believe.